xla/service/gpu/loop_double_buffer_transformer.cc

```diff
@@ -0,0 +1,415 @@
+    if (channel_id_comp_map.find(new_channel_id) == channel_id_comp_map.end()) {
+      channel_id_comp_map[new_channel_id] = new_instr->called_computations()[0];
+    } else {
+      channel_id_comp_map[new_channel_id]->AddAsyncInstruction(*new_instr);
+    }
+  } else if (hlo_query::IsCollectiveCommunicationOp(new_instr->opcode()) ||
+             hlo_query::IsAsyncCollectiveStartOp(new_instr->opcode())) {
+    new_instr->set_channel_id(hlo_query::NextChannelId(*module));
+  }
+}
+
+Status PeelInstructionsForOddTripCount(
+    HloModule* module, HloInstruction* while_instr,
+    const std::vector<HloInstruction*>& all_old_instructions,
+    const std::vector<HloInstruction*>& old_loop_roots,
+    const absl::flat_hash_map<HloInstruction*,
+                              std::vector<LoopInputConsumerInfo>>&
+        loop_input_consumers) {
+  HloCloneContext context(module, "peeled_double_buffer");
+
+  absl::flat_hash_map<HloInstruction*, HloInstruction*> old_to_new_map;
+  std::vector<HloInstruction*> all_new_instructions;
+  HloComputation* while_body = while_instr->while_body();
+  int64_t loop_output_count = while_body->root_instruction()->operand_count();
+  HloInstruction* input_parameter = while_body->parameter_instructions()[0];
+  HloInstruction* input_tuple = while_instr->mutable_operand(0);
+  CHECK(input_tuple->opcode() == HloOpcode::kTuple);
+
+  HloComputation* parent_comp = while_instr->parent();
+  std::vector<HloInstruction*> loop_inputs(loop_output_count);
+  for (int64_t i = 0; i < loop_output_count; i++) {
+    loop_inputs[i] = input_tuple->mutable_operand(i);
+  }
+  for (HloInstruction* old_instr : all_old_instructions) {
+    // This is to track mappings of old->new channel id for async
+    // collectives wrapped in the form of HloAsyncInstruction, the start and
+    // done need to have the same unique channel id.
+    if (old_instr == input_parameter || old_instr->IsUserOf(input_parameter)) {
+      continue;
+    }
+    VLOG(2) << "Peeling instruction " << old_instr->ToString();
+    HloInstruction* new_instr;
+    if (loop_input_consumers.find(old_instr) != loop_input_consumers.end()) {
+      // If instruction is the consumer of loop inputs, then we clone and put it
+      // in the map directly.
+      new_instr = parent_comp->AddInstruction(
+          old_instr->Clone(/*suffix=*/context.suffix()));
+      // There will be cases that other operands of this loop input consumer are
+      // already cloned, replace those operands with corresponding clones.
+      for (int64_t op_ind = 0; op_ind < new_instr->operand_count(); op_ind++) {
+        HloInstruction* op = new_instr->mutable_operand(op_ind);
+        if (old_to_new_map.find(op) != old_to_new_map.end()) {
+          TF_RETURN_IF_ERROR(
+              new_instr->ReplaceOperandWith(op_ind, old_to_new_map[op]));
+        }
+      }
+    } else {
+      // otherwise its operands should have been cloned already since
```

Comment:
I think we should pre-process all the "leaf" elements (the parameter(s) to the while loop) and add them to the map. Then, we wouldn't need to handle several cases in this copy loop.

This is an example of what I mean: https://github.com/openxla/xla/blob/main/xla/service/gpu/softmax_rewriter_triton.cc#L309. The whole copying logic is ~15 lines and has a single path through it, which I think would be more readable.